File: leetcode/0298-binary-tree-longest-consecutive-sequence/solution.cpp
#include <algorithm>
using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int longestConsecutive(TreeNode* root) {
        best = 0;
        dfs(root);
        return best;
    }

private:
    int best;

    int dfs(TreeNode* node) {
        if (node == nullptr) {
            return 0;
        }
        int length = 1;
        if (node->left != nullptr) {
            int leftLen = dfs(node->left);
            if (node->left->val == node->val + 1) {
                length = max(length, 1 + leftLen);
            }
        }
        if (node->right != nullptr) {
            int rightLen = dfs(node->right);
            if (node->right->val == node->val + 1) {
                length = max(length, 1 + rightLen);
            }
        }
        best = max(best, length);
        return length;
    }
};

File: leetcode/0298-binary-tree-longest-consecutive-sequence/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int best = 0;

    public int longestConsecutive(TreeNode root) {
        dfs(root);
        return best;
    }

    private int dfs(TreeNode node) {
        if (node == null) {
            return 0;
        }
        int length = 1;
        if (node.left != null) {
            int leftLen = dfs(node.left);
            if (node.left.val == node.val + 1) {
                length = Math.max(length, 1 + leftLen);
            }
        }
        if (node.right != null) {
            int rightLen = dfs(node.right);
            if (node.right.val == node.val + 1) {
                length = Math.max(length, 1 + rightLen);
            }
        }
        best = Math.max(best, length);
        return length;
    }
}
